lib/within.js

```diff
@@ -21,19 +21,68 @@ function within(context, fn) {
       if (helpers[helper]._withinBegin) recorder.add('start within block', () => helpers[helper]._withinBegin(context));
     });
 
-    fn();
 
-    Object.keys(helpers).forEach((helper) => {
-      if (helpers[helper]._withinEnd) recorder.add('finish within block', () => helpers[helper]._withinEnd());
-    });
-
-    recorder.add('restore session', () => {
-      recorder.session.restore();
+    try {
```

Comment:
This is copy-paste from scenario.js

I think we can reduce the code duplication. I think we can have a `resumeTest` exported from scenario.js.

```js
module.exports.resumeTest = function(done) {
          recorder.add('create new promises queue for generator', function (data) {
            recorder.session.start('generator'); // creating a new promise chain
            try {
              let resume = res.next(data);
              if (resume.done) {
                done();
              } else {
                resumeTest(done);
              }
            } catch (err) {
              recorder.throw(err); // catch error in test and rethrow into promise
            }
            recorder.catch(); // handle all errors in promise chain
          });
}
```

Usage inside scenario.js:

```js
     if (isGenerator(testFn)) {
        try {
          res.next(); // running test
        } catch (err) {
          event.emit(event.test.failed, test, err);
          done(err);
          return test;
        }
        recorder.catch(); // catching possible errors in promises
        resumeTest(() => {
                recorder.add('fire test.passed', () => event.emit(event.test.passed, test));
                recorder.add('finish generator with no error', () => done()); // finish him
        });
```

inside within:

```js
      if (isGenerator(testFn)) {
        try {
          res.next(); // running test
        } catch (err) {
          event.emit(event.test.failed, test, err);
          done(err);
          return test;
        }
        recorder.catch(); // catching possible errors in promises
        resumeTest(() => {

                Object.keys(helpers).forEach((helper) => {
                  if (helpers[helper]._withinEnd) recorder.add('finish within block', () => helpers[helper]._withinEnd());
                });
                recorder.add('Finalize session within_generator', () => {
                  recorder.session.restore('within_generator');
                });
      });
```